### PR TITLE
Feature Noise XKpsk3 integration connection

### PR DESCRIPTION
**Pull request by simonwicky**
# Description

This is a part of stacked PRs to integrate XKpsk3 Noise connection between nym nodes

This PR concerns Noise connection and fallback logic

The stack of PR is as follow :

#4360 -> #4361 -> #4373 -> #4362 (this one)

# Checklist:

- [ ] added a changelog entry to `CHANGELOG.md`

### CODE REVIEW

File: common/client-libs/mixnet-client/src/client.rs [hunk 1]
```diff
@@ -97,7 +115,41 @@ impl Client {
...
                     current_reconnection.store(0, Ordering::Release);
-                    Framed::new(stream, NymCodec)
+                    //Get the topology, because we need the keys for the handshake
+                    let topology_ref = match topology_access.current_topology().await {
+                        Some(topology) => topology,
+                        None => {
+                            error!("Cannot perform Noise handshake to {address}, due to topology error");
+                            return;
+                        }
+                    };
```
Comment (octol) [inline common/client-libs/mixnet-client/src/client.rs]:
```
let Some(topology) = ... else {
    return;
}
```

File: common/client-libs/mixnet-client/src/client.rs [hunk 2]
```diff
@@ -97,7 +115,41 @@ impl Client {
...
+
+                    let epoch_id = match api_client.get_current_epoch_id().await {
+                        Ok(id) => id,
+                        Err(err) => {
+                            error!("Cannot perform Noise handshake to {address}, due to epoch id error - {err}");
+                            return;
+                        }
+                    };
```
Comment (octol) [inline common/client-libs/mixnet-client/src/client.rs]: See comment about `let else`.
Another option is to make the function return `Result<()>` and just to `tap_err` + `?`.

File: common/client-libs/mixnet-client/src/client.rs [hunk 3]
```diff
@@ -97,7 +115,41 @@ impl Client {
...
+                        &local_identity.private_key().to_bytes(),
+                    )
+                    .await
+                    {
+                        Ok(noise_stream) => noise_stream,
+                        Err(err) => {
+                            error!("Failed to perform Noise handshake with {address} - {err}");
+                            return;
+                        }
+                    };
```
Comment (octol) [inline common/client-libs/mixnet-client/src/client.rs]: Similar to above

File: common/nymnoise/src/error.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum NoiseError {
+    #[error("encountered a Noise decryption error")]
+    DecryptionError,
+
+    #[error("encountered a Noise Protocol error - {0}")]
+    ProtocolError(Error),
+    #[error("encountered an IO error - {0}")]
```
Comment (octol) [inline common/nymnoise/src/error.rs:17]: nit: blank line

File: common/nymnoise/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    Ok(Connection::Noise(noise_stream.perform_handshake().await?))
+}
+
+pub async fn upgrade_noise_initiator_with_topology(
+    conn: TcpStream,
+    pattern: NoisePattern,
+    topology: &NymTopology,
+    epoch: u32,
+    local_public_key: &[u8],
+    local_private_key: &[u8],
```
Comment (octol) [inline common/nymnoise/src/lib.rs]: Maybe make these two strongly typed to prevent accidentally reversing them?

File: common/nymnoise/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+) -> Result<Connection, NoiseError> {
+    //Get init material
+    let responder_addr = match conn.peer_addr() {
+        Ok(addr) => addr,
+        Err(err) => {
+            error!("Unable to extract peer address from connection - {err}");
+            return Err(Error::Prereq(Prerequisite::RemotePublicKey).into());
+        }
+    };
```
Comment (octol) [inline common/nymnoise/src/lib.rs]: This looks like a `.map_err()?`

File: gateway/src/node/mixnet_handling/receiver/connection_handler.rs [hunk 7]
```diff
@@ -205,7 +222,41 @@ impl<St: Storage> ConnectionHandler<St> {
...
         shutdown.mark_as_success();
-        let mut framed_conn = Framed::new(conn, NymCodec);
+
+        let topology_ref = match self.topology_access.current_topology().await {
+            Some(topology) => topology,
+            None => {
+                error!("Cannot perform Noise handshake to {remote}, due to topology error");
+                return;
+            }
+        };
```
Comment (octol) [inline gateway/src/node/mixnet_handling/receiver/connection_handler.rs:229]: `let else`

File: gateway/src/node/mixnet_handling/receiver/connection_handler.rs [hunk 8]
```diff
@@ -205,7 +222,41 @@ impl<St: Storage> ConnectionHandler<St> {
...
+
+        let epoch_id = match self.api_client.get_current_epoch_id().await {
+            Ok(id) => id,
+            Err(err) => {
+                error!("Cannot perform Noise handshake to {remote}, due to epoch id error - {err}");
+                return;
+            }
+        };
```
Comment (octol) [inline gateway/src/node/mixnet_handling/receiver/connection_handler.rs:237]: `let else` or `?` + return type change

File: gateway/src/node/mixnet_handling/receiver/connection_handler.rs [hunk 9]
```diff
@@ -205,7 +222,41 @@ impl<St: Storage> ConnectionHandler<St> {
...
+            &self.local_identity.public_key().to_bytes(),
+            &self.local_identity.private_key().to_bytes(),
+        )
+        .await
+        {
+            Ok(noise_stream) => noise_stream,
+            Err(err) => {
+                error!("Failed to perform Noise handshake with {remote} - {err}");
+                return;
+            }
```
Comment (octol) [inline gateway/src/node/mixnet_handling/receiver/connection_handler.rs:253]: `let else`

File: mixnode/src/node/listener/connection_handler/mod.rs [hunk 10]
```diff
@@ -88,8 +102,43 @@ impl ConnectionHandler {
...
         shutdown.mark_as_success();
-        let mut framed_conn = Framed::new(conn, NymCodec);
+
+        let topology_ref = match self.topology_access.current_topology().await {
+            Some(topology) => topology,
+            None => {
+                error!("Cannot perform Noise handshake to {remote}, due to topology error");
+                return;
+            }
+        };
```
Comment (octol) [inline mixnode/src/node/listener/connection_handler/mod.rs:111]: `let else`

File: mixnode/src/node/listener/connection_handler/mod.rs [hunk 11]
```diff
@@ -88,8 +102,43 @@ impl ConnectionHandler {
...
+
+        let epoch_id = match self.api_client.get_current_epoch_id().await {
+            Ok(id) => id,
+            Err(err) => {
+                error!("Cannot perform Noise handshake to {remote}, due to epoch id error - {err}");
+                return;
+            }
+        };
```
Comment (octol) [inline mixnode/src/node/listener/connection_handler/mod.rs:119]: `let else`

File: mixnode/src/node/listener/connection_handler/mod.rs [hunk 12]
```diff
@@ -88,8 +102,43 @@ impl ConnectionHandler {
...
+            &self.local_identity.private_key().to_bytes(),
+        )
+        .await
+        {
+            Ok(noise_stream) => noise_stream,
+            Err(err) => {
+                error!("Failed to perform Noise handshake with {remote} - {err}");
+                return;
+            }
+        };
```
Comment (octol) [inline mixnode/src/node/listener/connection_handler/mod.rs:136]: `let else`

**Review by octol — APPROVED** [review]

File: common/client-libs/mixnet-client/src/client.rs (hunk 1 above)
Comment (simonwicky) [reply common/client-libs/mixnet-client/src/client.rs]: Applied in bc64e28

File: common/client-libs/mixnet-client/src/client.rs (hunk 2 above)
Comment (simonwicky) [reply common/client-libs/mixnet-client/src/client.rs]: Changing return type causes more problem, and I want to keep the message error

File: common/client-libs/mixnet-client/src/client.rs (hunk 3 above)
Comment (simonwicky) [reply common/client-libs/mixnet-client/src/client.rs]: Changing return type causes more problem, and I want to keep the message error

File: common/nymnoise/src/error.rs (hunk 4 above)
Comment (simonwicky) [reply common/nymnoise/src/error.rs:17]: Applied in bc64e28

File: common/nymnoise/src/lib.rs (hunk 6 above)
Comment (simonwicky) [reply common/nymnoise/src/lib.rs]: Applied in bc64e28

File: gateway/src/node/mixnet_handling/receiver/connection_handler.rs (hunk 8 above)
Comment (simonwicky) [reply gateway/src/node/mixnet_handling/receiver/connection_handler.rs:237]: Changing return type causes more problem, and I want to keep the message error

File: gateway/src/node/mixnet_handling/receiver/connection_handler.rs (hunk 9 above)
Comment (simonwicky) [reply gateway/src/node/mixnet_handling/receiver/connection_handler.rs:253]: Changing return type causes more problem, and I want to keep the message error

File: gateway/src/node/mixnet_handling/receiver/connection_handler.rs (hunk 7 above)
Comment (simonwicky) [reply gateway/src/node/mixnet_handling/receiver/connection_handler.rs:229]: Applied in bc64e28

File: mixnode/src/node/listener/connection_handler/mod.rs (hunk 10 above)
Comment (simonwicky) [reply mixnode/src/node/listener/connection_handler/mod.rs:111]: Applied in bc64e28

File: mixnode/src/node/listener/connection_handler/mod.rs (hunk 11 above)
Comment (simonwicky) [reply mixnode/src/node/listener/connection_handler/mod.rs:119]: Changing return type causes more problem, and I want to keep the message error

File: mixnode/src/node/listener/connection_handler/mod.rs (hunk 12 above)
Comment (simonwicky) [reply mixnode/src/node/listener/connection_handler/mod.rs:136]: Changing return type causes more problem, and I want to keep the message error

File: common/nymnoise/src/lib.rs (hunk 5 above)
Comment (simonwicky) [reply common/nymnoise/src/lib.rs]: Applied in d2a9fa9

**Pull request merged**